File: frontend/src/types/assessment.ts
export interface AssessmentResponse {
  extracted_text: string
  cleaned_text: string
  tfidf_score: number
  sbert_score: number
  similarity_score: number
  keyword_overlap: number
  missing_keywords: string[]
  marks: number
  max_marks: number
  grade: string
  feedback: string
  assessed_at: string
}

export interface QuestionResult {
  question_id: string
  question: string
  student_answer: string
  model_answer: string
  similarity_score: number
  tfidf_score: number
  sbert_score: number
  keyword_overlap: number
  missing_keywords: string[]
  marks: number
  max_marks: number
  grade: string
  feedback: string
  status: 'completed' | 'failed'
  failure_reason?: string | null
}

export interface MultiAssessmentResponse {
  total_score: number
  max_total_score: number
  total_questions: number
  failed_questions: number
  status: 'completed' | 'partial_failure'
  student_id: string
  assessed_at: string
  results: QuestionResult[]
}

export type AnyAssessmentResult = AssessmentResponse | MultiAssessmentResponse

export function isMultiAssessment(result: AnyAssessmentResult): result is MultiAssessmentResponse {
  return 'results' in result && 'total_questions' in result
}

export interface HealthStatus {
  status: 'healthy' | 'degraded'
  ocr_available: boolean
  tesseract_available: boolean
  vision_credentials_configured: boolean
  sbert_loaded: boolean
  spacy_model_loaded: boolean
  llm_configured: boolean
  pdf_support: boolean
  capabilities: {
    ocr: boolean
    semantic_similarity: boolean
    llm: boolean
    pdf: boolean
  }
  supported_modes: {
    manual_assessment: boolean
    auto_reference_generation: boolean
    multi_question_assessment: boolean
  }
  timestamp: string
}

export interface SingleHistoryResult {
  id: string
  result_type: 'single_question'
  student_id: string
  assessed_at: string
  question_id: string
  score_ratio: number
  marks: number
  max_marks: number
  grade: string
}

export interface MultiHistoryResult {
  id: string
  result_type: 'multi_question'
  student_id: string
  assessed_at: string
  total_questions: number
  total_score: number
  max_total_score: number
  average_score_ratio: number
}

export type HistoryResult = SingleHistoryResult | MultiHistoryResult

export interface ResultsResponse {
  results: HistoryResult[]
  count: number
}

// ── Progress tracking types ──

export interface ProgressEvent {
  task_id: string
  stage: string
  status: 'running' | 'completed' | 'error'
  message: string
  step: number
  total_steps: number
  completed_stages: string[]
}

export interface TaskStartResponse {
  task_id: string
  status: 'processing'
}

export const PIPELINE_STAGES = [
  { key: 'upload_received', label: 'Upload received', icon: 'Upload' },
  { key: 'file_type_detection', label: 'File type detection', icon: 'FileSearch' },
  { key: 'text_extraction', label: 'Text extraction', icon: 'ScanText' },
  { key: 'nlp_preprocessing', label: 'NLP preprocessing', icon: 'Braces' },
  { key: 'question_detection', label: 'Question detection', icon: 'Search' },
  { key: 'llm_generation', label: 'AI answer generation', icon: 'Sparkles' },
  { key: 'answer_mapping', label: 'Answer mapping', icon: 'ArrowRightLeft' },
  { key: 'similarity', label: 'Similarity computation', icon: 'GitCompare' },
  { key: 'scoring', label: 'Scoring & feedback', icon: 'Award' },
  { key: 'completed', label: 'Complete', icon: 'CheckCircle' },
] as const

File: backend/models/assessment.py
"""Pydantic schemas for assessment request/response validation."""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class AssessmentBaseModel(BaseModel):
    """Shared base model for assessment schemas."""

    model_config = ConfigDict(protected_namespaces=())


class AssessmentResponse(AssessmentBaseModel):
    """JSON response shape for POST /api/assess (single-question, backward compat)."""

    extracted_text: str = Field(description="Raw OCR output from Google Vision")
    cleaned_text: str = Field(description="NLP-preprocessed student text")
    tfidf_score: float = Field(ge=0.0, le=1.0)
    sbert_score: float = Field(ge=-1.0, le=1.0)
    similarity_score: float = Field(description="Weighted combined score")
    keyword_overlap: float = Field(ge=0.0, le=1.0)
    missing_keywords: list[str] = Field(default_factory=list)
    marks: float = Field(ge=0.0)
    max_marks: int = Field(ge=1)
    grade: str = Field(description="Letter grade: A/B/C/D/F")
    feedback: str
    assessed_at: str = Field(description="ISO 8601 timestamp")


class QuestionResult(AssessmentBaseModel):
    """Result for a single question within a multi-question assessment."""

    question_id: str
    question: str = Field(default="", description="Detected question text")
    student_answer: str = Field(description="Student's answer for this question")
    model_answer: str = Field(description="LLM-generated or provided model answer")
    similarity_score: float = Field(description="Weighted combined score")
    tfidf_score: float = Field(ge=0.0, le=1.0)
    sbert_score: float = Field(ge=-1.0, le=1.0)
    sentence_similarity: float = Field(default=0.0, ge=0.0, le=1.0)
    concept_coverage: float = Field(default=0.0, ge=0.0, le=1.0)
    entailment_score: float | None = Field(default=None)
    keyword_overlap: float = Field(ge=0.0, le=1.0)
    missing_keywords: list[str] = Field(default_factory=list)
    marks: float = Field(ge=0.0)
    max_marks: int = Field(ge=1)
    grade: str
    feedback: str
    status: Literal["completed", "failed"] = "completed"
    failure_reason: str | None = None


class MultiAssessmentResponse(AssessmentBaseModel):
    """JSON response shape for multi-question assessment."""

    total_score: float = Field(ge=0.0)
    max_total_score: int = Field(ge=1)
    total_questions: int = Field(ge=1)
    failed_questions: int = Field(ge=0)
    status: Literal["completed", "partial_failure"] = "completed"
    student_id: str
    assessed_at: str = Field(description="ISO 8601 timestamp")
    results: list[QuestionResult]


class TaskStartResponse(AssessmentBaseModel):
    """Accepted async assessment task."""

    task_id: str
    status: Literal["processing"]


class SingleHistoryEntry(AssessmentBaseModel):
    """History-safe summary for a single-question assessment."""

    id: str
    result_type: Literal["single_question"]
    student_id: str
    assessed_at: str
    question_id: str
    score_ratio: float = Field(ge=0.0, le=1.0)
    marks: float = Field(ge=0.0)
    max_marks: int = Field(ge=1)
    grade: str


class MultiHistoryEntry(AssessmentBaseModel):
    """History-safe summary for an aggregate multi-question assessment."""

    id: str
    result_type: Literal["multi_question"]
    student_id: str
    assessed_at: str
    total_questions: int = Field(ge=1)
    total_score: float = Field(ge=0.0)
    max_total_score: int = Field(ge=1)
    average_score_ratio: float = Field(ge=0.0, le=1.0)


class ErrorResponse(AssessmentBaseModel):
    """Standard error envelope."""

    error: str
    code: int
